src/base/QXmppExternalServiceDiscoveryIq.cpp

```diff
@@ -0,0 +1,483 @@
+///
+std::optional<QString> QXmppExternalService::name() const
+{
+    return d->name;
+}
+
+///
+/// Sets the name of the external service.
+///
+void QXmppExternalService::setName(const QString &name)
+{
+    d->name = name;
+}
+
+///
+/// Returns the password of the external service.
+///
+std::optional<QString> QXmppExternalService::password() const
+{
+    return d->password;
+}
+
+///
+/// Sets the password of the external service.
+///
+void QXmppExternalService::setPassword(const QString &password)
+{
+    d->password = password;
+}
+
+///
+/// Returns the port of the external service.
+///
+std::optional<int> QXmppExternalService::port() const
+{
+    return d->port;
+}
+
+///
+/// Sets the port of the external service.
+///
+void QXmppExternalService::setPort(const int port)
+{
+    d->port = port;
+}
+
+///
+/// Returns the restricted mode of the external service.
+///
+std::optional<bool> QXmppExternalService::restricted() const
+{
+    return d->restricted;
+}
+///
+/// Sets the restricted mode of the external service.
+///
+void QXmppExternalService::setRestricted(const bool restricted)
+{
+    d->restricted = restricted;
+}
+
+///
+/// Returns the transport type of the external service.
+///
+std::optional<QXmppExternalService::Transport> QXmppExternalService::transport() const
+{
+    return d->transport;
+}
+
+///
+/// Sets the transport type of the external service.
+///
+void QXmppExternalService::setTransport(const Transport transport)
+{
+    d->transport = transport;
+}
+
+///
+/// Returns the username of the external service.
+///
+std::optional<QString> QXmppExternalService::username() const
+{
+    return d->username;
+}
+
+///
+/// Sets the username of the external service.
+///
+void QXmppExternalService::setUsername(const QString &username)
+{
+    d->username = username;
+}
+
+///
+/// Returns true if the element is a valid external service and can be parsed.
+///
+bool QXmppExternalService::isExternalService(const QDomElement &element)
+{
+    if (element.tagName() != "service") {
+        return false;
+    }
+
+    return element.hasAttribute("host") && !element.attribute("host").isEmpty() &&
+        element.hasAttribute("type") && !element.attribute("type").isEmpty();
+}
+
+///
+/// Parses given DOM element as an external service.
+///
+void QXmppExternalService::parse(const QDomElement &el)
+{
+    const QDomNamedNodeMap &attributes { el.attributes() };
+
+    setHost(el.attribute("host"));
+    setType(el.attribute("type"));
+
+    if (attributes.contains("action")) {
+        const std::optional<Action> action { actionFromString(el.attribute("action")) };
+
+        if (action.has_value()) {
+            setAction(action.value());
+        }
+    }
+
+    if (attributes.contains("expires")) {
+        const QString &expiresAttribute { el.attribute("expires") };
+
+        QDateTime expires;
+        Qt::DateFormat dateFormat {
+            expiresAttribute.contains(QRegExp("\\.[0-9]{3}"))
```

Comment:
QRegExp is replaced by QRegularExpression in Qt6. Would it be possible to use that instead?

Reply:
Fixed.